tool/planet/agent.go

```diff
@@ -232,27 +254,95 @@ var electedUnits = []string{
 	"kube-apiserver.service",
 }
 
-func unitsCommand(command string) error {
-	log.Debugf("executing %v on %v", command, electedUnits)
-	var errors []error
-	for _, unit := range electedUnits {
-		cmd := exec.Command("/bin/systemctl", command, unit)
-		log.Debugf("executing %v", cmd)
-		out, err := cmd.CombinedOutput()
-		if err != nil {
-			errors = append(errors, err)
-			// Instead of failing immediately, complete start of other units
-			log.Warningf("failed to execute %v: %s\n%v", cmd, out, trace.DebugReport(err))
+func newUnitMonitor() (*unitMonitor, error) {
+	conn, err := systemdDbus.New()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return &unitMonitor{
+		conn:  conn,
+		units: electedUnits,
+	}, nil
+}
+
+type unitMonitor struct {
+	conn  *systemdDbus.Conn
+	units []string
+}
+
+func (r *unitMonitor) close() {
+	r.conn.Close()
+}
+
+func (r *unitMonitor) start(ctx context.Context) {
+	stopC := make(chan string, len(r.units))
+	// Number of successfully started start jobs
+	var numStarted int
+	for _, unit := range r.units {
+		logger := log.WithField("unit", unit)
+		logger.Info("Start service.")
+		if _, err := r.conn.StartUnit(unit, "replace", stopC); err != nil {
+			logger.WithError(err).Warn("Failed to start unit.")
+			continue
+		}
+		numStarted += 1
+	}
+	if numStarted == 0 {
+		return
+	}
+	r.waitForJobs(ctx, stopC, numStarted)
+}
+
+func (r *unitMonitor) stop(ctx context.Context) error {
+	stopC := make(chan string, len(r.units))
+	// Number of successfully started stop jobs
+	var numStopped int
+	for _, unit := range r.units {
+		logger := log.WithField("unit", unit)
+		logger.Info("Shut down service.")
+		if _, err := r.conn.StopUnit(unit, "replace", stopC); err != nil {
+			log.WithError(err).WithField("unit", unit).Warn("Failed to stop unit.")
+			continue
+		}
+		numStopped += 1
+	}
+	if numStopped != 0 {
+		r.waitForJobs(ctx, stopC, numStopped)
+	}
+	// NB: ListUnitsByPatterns is not supported on systemd 219
+	failedUnits, err := r.conn.ListUnitsFiltered([]string{"failed"})
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	for _, unit := range failedUnits {
+		if err := r.conn.ResetFailedUnit(unit.Name); err != nil {
+			log.WithError(err).WithField("unit", unit.Name).Warn("Failed to reset failed unit.")
+		}
+	}
+	return nil
+}
+
+func (r *unitMonitor) waitForJobs(ctx context.Context, stopC <-chan string, services int) {
```

Comment:
This looks like the same could be achieved by using wait groups from stdlib.

Reply:
Which aspect of this exactly do you mean to replace with waitgroups?
`stopC` channel goes into the `go-systemd` library which will be sending on it.